### PR TITLE
[FIR] fix mismatching type parameter symbols in generic class

### PR DESCRIPTION
**Pull request by juan-chen**
Type parameters in generic classes are given wrapped descriptors
whereas references to those type parameters inside the classes have
deserialized descriptors (from stub generation). This is a problem for
creating type substitutions for instantiated classes. This commit
fixes the mismatch by using deserialized descriptors if possible.

### CODE REVIEW

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt [hunk 1]
```diff
@@ -282,11 +316,12 @@ class Fir2IrClassifierStorage(
     private fun createIrTypeParameterWithoutBounds(
         typeParameter: FirTypeParameter,
         index: Int,
-        typeContext: ConversionTypeContext = ConversionTypeContext.DEFAULT
+        typeContext: ConversionTypeContext = ConversionTypeContext.DEFAULT,
+        existingDescriptor: TypeParameterDescriptor? = null
     ): IrTypeParameter {
         require(index >= 0)
-        val descriptor = WrappedTypeParameterDescriptor()
         val origin = IrDeclarationOrigin.DEFINED
```
Comment (jsjeon) [inline compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt:323]: For those new handled case(s), I wonder the origin of type parameter, which would be the same as that of the enclosing class, i.e., external stub, should be propagated and recorded in that way. How about moving this `origin` to the param of `createIrTypeParameterWithoutBounds` with default origin `DEFINED` and pass that external stub if necessary?

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt [hunk 2]
```diff
@@ -113,11 +115,43 @@ class Fir2IrClassifierStorage(
 
     private fun IrClass.declareTypeParameters(klass: FirClass<*>) {
         if (klass is FirRegularClass) {
-            preCacheTypeParameters(klass)
+            if (origin == IrDeclarationOrigin.IR_EXTERNAL_DECLARATION_STUB
```
Comment (jsjeon) [inline compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt:118]: There is external Java stub as well. Should we handle it in a similar way or not necessary?

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt (hunk 1 above)
Comment (juan-chen) [reply compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt]: done

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt (hunk 2 above)
Comment (juan-chen) [reply compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt]: done

**Comment by mglukhikh** [comment]
It will be probably easier if I think about getting symbol by signature myself. But think please about some test confirming this behavior.

**Comment by mglukhikh** [comment]
NB: continued in #3480

**Pull request closed without merging**